### DISCUSSION TITLE
Percent encoding of package name

### DISCUSSION THREAD

**Discussion by kburich** [package-url/purl-spec#282] [Q&A] [answered]
Not sure if I am understanding the spec correctly regarding percent encoding. Is **pkg:npm/foo!** a valid purl?
If so, does it represent same package as purl **pkg:npm/foo%21**?

**Comment by matt-phylum** [comment, marked as answer]
Both are valid and are supposed to represent the same package.

althonos/packageurl.rs, giterlizzi/perl-URI-PackageURL, maennchen/purl, package-url/packageurl-dotnet, package-url/packageurl-go, package-url/packageurl-java, package-url/packageurl-js, package-url/packageurl-php, package-url/packageurl-python, package-url/packageurl-ruby, package-url/packageurl-swift, phylum-dev/purl, sonatype/package-url-java (all implementations I know of) read `pkg:npm/foo!` as having name `foo!`.

althonos/packageurl.rs, maennchen/purl, package-url/packageurl-dotnet, package-url/packageurl-js, package-url/packageurl-swift, phylum-dev/purl (6/13) format an NPM package `foo!` as `pkg:npm/foo!`.

giterlizzi/perl-URI-PackageURL, package-url/packageurl-go, package-url/packageurl-java, package-url/packageurl-php, package-url/packageurl-python, package-url/packageurl-ruby, sonatype/package-url-java (7/13) format an NPM package `foo!` as `pkg:npm/foo%21`.

All implementations except maennchen/purl (https://github.com/maennchen/purl/issues/10) read `pkg:npm/foo%21` as having name `foo!`, the same as `pkg:npm/foo!`.

**Reply by kburich** [reply]
Hmm, does this mean that purl `pkg:npm/foo%2521` represents the same package as purl `pkg:npm/foo%21` since percent encoding for `%` is `%25`?

**Reply by matt-phylum** [reply]
No. `pkg:npm/foo%2521` represents a package named `foo%21`. maennchen/purl gets it wrong for the same reason as before, but all other implementations agree.

**Reply by kburich** [reply]
Ah ok I get it now, thanks.
